More NT subclass op support for SAM

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111253

With this PR, we have full op support for SAM without needing to unwrap subclass into jagged buffer -> run ops -> rewrap manually. Specifically, this was previously happening in the MaskDecoder.

cc @cpuhrsch @bhosmer @drisspg @soulitzer @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

jbschlosser committed:
More NT subclass op support for SAM

[ghstack-poisoned]

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -130,22 +137,59 @@ def jagged_unary_pointwise(func, *args, **kwargs):
+    a_is_nt = isinstance(a, NestedTensor)
+    extracted_kwargs = extract_kwargs(a) if a_is_nt else extract_kwargs(b)
+
+    # === Handle broadcasting across the batch / ragged dims ===
```
no action needed, but eventually we'd need autograd support for implicit broadcasting as HSTU folks have found out

jbschlosser committed:
Update on "More NT subclass op support for SAM"


With this PR, we have full op support for SAM without needing to unwrap subclass into jagged buffer -> run ops -> rewrap manually. Specifically, this was previously happening in the MaskDecoder.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jbschlosser committed:
Update on "More NT subclass op support for SAM"


With this PR, we have full op support for SAM without needing to unwrap subclass into jagged buffer -> run ops -> rewrap manually. Specifically, this was previously happening in the MaskDecoder.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

soulitzer commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -161,6 +160,27 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+            output = func(*t_args, **t_kwargs)
+            return NestedTensor(output, **extract_kwargs(args[0]))
+        with torch._C.DisableTorchFunctionSubclass():
+            return func(*args, **kwargs)
```
What's the plan here with torch compile?

soulitzer replied:
(I suspect tests would fail)

jbschlosser replied:
Yeah good question, I need to look into this. For the purposes of getting SAM running without torch.compile, this is fine, but not great for the immediate next step

jbschlosser committed:
Update on "More NT subclass op support for SAM"


With this PR, we have full op support for SAM without needing to unwrap subclass into jagged buffer -> run ops -> rewrap manually. Specifically, this was previously happening in the MaskDecoder.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jbschlosser committed:
Update on "More NT subclass op support for SAM"


With this PR, we have full op support for SAM without needing to unwrap subclass into jagged buffer -> run ops -> rewrap manually. Specifically, this was previously happening in the MaskDecoder.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jbschlosser committed:
Update on "More NT subclass op support for SAM"


With this PR, we have full op support for SAM without needing to unwrap subclass into jagged buffer -> run ops -> rewrap manually. Specifically, this was previously happening in the MaskDecoder.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jbschlosser commented on torch/nested/_internal/ops.py:
```diff
@@ -215,10 +309,7 @@ def linear_default(func, *args, **kwargs):
-    if bias is not None:
-        new_values += bias
-    return NestedTensor(new_values, **extract_kwargs(inp))
+    return NestedTensor(func(inp._values, **new_kwargs), **extract_kwargs(inp))
```
@soulitzer FYI this change fixes linear to expect `weight` in the form of `(out_channels, in_channels)`. Backward and tests have to change to accommodate this as well

soulitzer replied:
whoops, thanks!

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "More NT subclass op support for SAM"


With this PR, we have full op support for SAM without needing to unwrap subclass into jagged buffer -> run ops -> rewrap manually. Specifically, this was previously happening in the MaskDecoder.

cc cpuhrsch bhosmer drisspg soulitzer jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge -f "ignore spurious failure"

huydhn commented:
@pytorchbot drci

(please ignore this, I'm testing Dr.CI)

bdhirsh commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -742,6 +742,10 @@ static std::tuple<Tensor, bool> batchify(
     const Tensor& input,
     const int64_t num_spatial_dims,
     const std::string& func_name) {
+  // assume NTs are always batched
+  if (input.is_nested()) {
+    return std::make_tuple(input, true);
+  }
```
just to give an alternative: You could override `conv2d_transpose` for the `AutogradNestedTensor` key using `py_impl`. But maybe this is more of a pain, since it would require you to re-implement more of `conv2d_transpose` in python?

jbschlosser replied:
ah thanks for the suggestion! Indeed this would be more work, as I'd have to implement more of `conv2d_transpose`, alongside the other 1d/2d/3d transposed / non-transposed variants.
